src/Metrics/Metric.h

```diff
@@ -2,60 +2,52 @@
 
 #include <atomic>
 #include <cstdint>
+#include <memory>
 #include <string>
 
 namespace Metrics {
 
-class ProcessorInterface;
 class Registrar;
 
-enum class Severity { DEBUG, INFO, WARNING, ERROR };
+enum struct Severity { DEBUG, INFO, WARNING, ERROR };
 
-using CounterType = std::atomic<std::int64_t>;
-using InternalCounterType = decltype(((CounterType *)(nullptr))->load());
+using CounterType = std::atomic<int64_t>;
 
 class Metric {
 public:
-  Metric(std::string Name, std::string Description,
-         Severity Level = Severity::DEBUG)
-      : MName(std::move(Name)), MDesc(std::move(Description)), SevLvl(Level) {}
+  Metric(std::shared_ptr<Registrar> RegistrarUsed, std::string Name,
```

Comment:
Providing a registrar to a Metrics::Metric constructor is bad interface design if we are also requiring the user to call Registrar::registerMetric. Have one alternative, not both.
Furthermore, the current implementation requires us to (indirectly or directly) hold on to the registrar after we are done with it.

Reply:
Agree on the first point, should be easily resolved.
The second point is true, and longer object lifetime is generally bad. I'll have a think about it.